lib/src/migrators/namespace.dart

```diff
@@ -0,0 +1,245 @@
+      : super(importCache, migrateDependencies);
+
+  @override
+  void visitStylesheet(Stylesheet node) {
+    var oldSpansByNamespace = _spansByNamespace;
+    var oldUsedNamespaces = _usedNamespaces;
+    _spansByNamespace = {};
+    _usedNamespaces = {};
+    super.visitStylesheet(node);
+    _spansByNamespace = oldSpansByNamespace;
+    _usedNamespaces = oldUsedNamespaces;
+  }
+
+  @override
+  void beforePatch(Stylesheet node) {
+    // Pass each `@use` rule through
+    var newNamespaces = <String, Set<UseRule>>{};
+    for (var rule in node.children.whereType<UseRule>()) {
+      if (rule.namespace == null) continue;
+      newNamespaces
+          .putIfAbsent(renamer.rename(rule) ?? rule.namespace, () => {})
+          .add(rule);
+    }
+    var unreferencedRules = <UseRule>{};
+    for (var entry in newNamespaces.entries) {
+      var newNamespace = entry.key;
+      var rules = entry.value;
+      if (rules.length == 1) {
+        if (unreferencedFlag == UnreferencedFlag.all &&
+            !_spansByNamespace.containsKey(rules.first.namespace)) {
+          unreferencedRules.add(rules.first);
+        } else {
+          _patchNamespace(rules.first, newNamespace);
+        }
+        continue;
+      }
+      if (unreferencedFlag != UnreferencedFlag.none) {
+        for (var rule in rules.toSet()) {
+          if (!_spansByNamespace.containsKey(rule.namespace)) {
+            unreferencedRules.add(rule);
+            rules.remove(rule);
+          }
+        }
+      }
+      if (rules.length == 1) {
+        _patchNamespace(rules.first, newNamespace);
+        continue;
+      }
+      if (!forceRename) {
+        throw MigrationException(
+            'Rename failed. ${rules.length} rules would use namespace '
+            '"$newNamespace".\n'
+            'Run with --force to rename with numerical suffixes.');
+      }
+      var suffix = 2;
+      for (var rule in rules) {
+        var forcedNamespace = newNamespace;
+        while (_usedNamespaces.contains(forcedNamespace)) {
+          forcedNamespace = '$newNamespace$suffix';
+          suffix++;
+        }
+        _patchNamespace(rule, forcedNamespace);
+      }
+    }
+    var suffix = 1;
+    for (var rule in unreferencedRules) {
+      var namespace = '_unreferenced$suffix';
+      while (_usedNamespaces.contains(namespace)) {
+        namespace = '_unreferenced$suffix';
+        suffix++;
+      }
+      _patchNamespace(rule, namespace);
+    }
+  }
+
+  /// Patch [rule] and all references to it with [newNamespace].
+  void _patchNamespace(UseRule rule, String newNamespace) {
+    _usedNamespaces.add(newNamespace);
+    if (rule.namespace == newNamespace) return;
+    var asClause =
+        RegExp('\\sas\\s(${rule.namespace})').firstMatch(rule.span.text);
+    if (asClause == null) {
+      // Add an `as` clause to a rule that previously lacked one.
+      var end = RegExp(r"""@use\s("|').*\1""").firstMatch(rule.span.text).end;
```

Comment:
```suggestion
      var end = RegExp(r"""@use\s("|').*?\1""").firstMatch(rule.span.text).end;
```

If the `.*` here is greedy, it would match all the way into a `with` clause that contained quoted string arguments, which is definitely not what we want.